src/Environment.jl

```diff
@@ -0,0 +1,44 @@
+
+const _NREFS = Ref(0)
+const _INITIALIZED = Ref(false)
+
+function Init(parts::MPIData;p4est_verbosity_level=P4est_wrapper.SC_LP_DEFAULT,finalize_atexit=true)
+  if !MPI.Initialized()
+      MPI.Init()
+  end
+
+  if finalize_atexit
+    atexit(Finalize)
+  end
+  Finalize()
+
+  # sc_init(parts.comm, Cint(true), Cint(true), C_NULL, p4est_verbosity_level)
+  p4est_init(C_NULL, p4est_verbosity_level)
+  _INITIALIZED[] = true
+
+  return nothing
+end
+
+function Initialized()
+  return _INITIALIZED[] == true
+end
+
+function Finalize()
+  if Initialized()
+    GC.gc() # Finalize all object out of scope at this point
+    if _NREFS[] != 0
+      @warn "$(_NREFS[]) objects still not finalized before calling GridapP4est.Finalize()"
+    end
+    _NREFS[] = 0
+    _INITIALIZED[] = false
+    #sc_finalize()
+  end
+  return nothing
+end
+
+function with(f,parts::MPIData;kwargs...)
+  Init(parts;kwargs...)
+  out = f()
```

Comment:
what if `f` (user-defined function, right?) needs args + kwargs?

Reply:
Usually `f` comes from a `do...end` block, which does not have args nor kwargs. This is how `GridapPETSc` does it.

Reply:
Yes. The philosophy here is that All the input f needs is captured from the surrounding environment (f becomes a closure). I was asking this because the prun method from partitioned arrays didn't have args and kwargs in a first version (now in the latest version they have) and that was leading to lack of expressivity to some extent in some situations. However, i thin that the with method is somehow different as you typically call it from the main method being triggered from prun so that all data f needs is already in local scope.

Reply:
Let's proceed wo args and kwargs